New Resource : `azurerm_api_management_gateway_config_connection`

sinbai opened the pull request:
<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->

Swagger: https://github.com/Azure/azure-rest-api-specs/blob/a61456ebadb3ad73cee898b8bdd2f14885663332/specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2024-05-01/apimgatewayConfigConnections.json#L130
## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevant documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.


## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->

Test results in TC:
<img width="581" height="488" alt="image" src="https://github.com/user-attachments/assets/7ca69c1b-42c4-4d1f-b1a1-740a9d0ad713" />

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `New Resource` - `azurerm_api_management_gateway_config_connection`


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #30522

## Rollback Plan

If a change needs to be reverted, we will publish an updated version of the provider.

## Changes to Security Controls

Are there any changes to security controls (access controls, encryption, logging) in this pull request? If so, explain.

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

sinbai committed:
support new resource azurerm_api_management_gateway_config_connection

ms-zhenhua reviewed (changes requested):
Hi @sinbai ,

Thanks for this PR - I've taken a look through and left some comments inline. If we can fix those up, this should be good to go 👍

sinbai committed:
fix comments

sinbai commented on website/docs/r/api_management_gateway_config_connection.html.markdown:
```diff
@@ -0,0 +1,93 @@
+
+* `name` - (Required) The name of the API Management Gateway Config Connection. Changing this forces a new resource to be created.
+
+* `api_management_gateway_id` - (Required) The ID of the API management gateway that this connection belongs to. Changing this forces a new resource to be created.
```
Fixed.

sinbai commented:
> Hi @sinbai ,
> 
> Thanks for this PR - I've taken a look through and left some comments inline. If we can fix those up, this should be good to go 👍

Hi @ms-zhenhua thanks for your feedback. I've addressed the comments, and I'd appreciate it if you could take another look?

ms-zhenhua reviewed (approved):
LGTM~

catriona-m reviewed (changes requested):
HI @sinbai thanks for working on this. I think this is looking good, I just left one query inline for you to consider. Thanks!

teowa committed:
Merge branch 'main' of github.com:hashicorp/terraform-provider-azurerm into apim/new_resource_workspace_gateway_connection

catriona-m commented on internal/services/apimanagement/api_management_gateway_config_connection_resource.go:
```diff
@@ -0,0 +1,181 @@
+
+		"api_management_gateway_id": commonschema.ResourceIDReferenceRequiredForceNew(&apigateway.GatewayId{}),
+
+		"source_id": commonschema.ResourceIDReferenceRequiredForceNew(&workspace.WorkspaceId{}),
```
would it make sense for this to be called `api_management_workspace_id` instead?

teowa replied:
Updated to `api_management_workspace_id`. I believe the resource name should remain unchanged, since the resource type is `Microsoft.ApiManagement/gateways/configConnections`, and in the future, there may be additional source IDs beyond just the workspace ID, and we can support it by adding a new _id field in this resource. WDYT?

catriona-m reviewed (changes requested):
Thanks @sinbai, I just left one comment inline about the `source_id` property - this only ever seems to be an `api_management_workspace_id`, so should we consider changing the name to be more reflective of this? Thanks!

teowa committed:
Merge branch 'main' of github.com:hashicorp/terraform-provider-azurerm into apim/new_resource_workspace_gateway_connection

teowa committed:
rename source_id to api_management_workspace_id
